src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1309,85 +1311,206 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
+                            {
+                                goto NotEqual;
+                            }
+                            offset += (nuint)Vector256<byte>.Count;
+                        } while (lengthToExamine > offset);
                     }
-                    return LoadVector(ref first, lengthToExamine) == LoadVector(ref second, lengthToExamine);
-                }
 
-                Debug.Assert((byte*)lengthToExamine >= (byte*)sizeof(UIntPtr));
+                    // Do final compare as Vector256<byte>.Count from end rather than start
+                    Debug.Assert(lengthToExamine >= 0);
+                    vecResult = Avx2.CompareEqual(LoadVector256(ref first, lengthToExamine), LoadVector256(ref second, lengthToExamine));
+                    if (Avx2.MoveMask(vecResult) == -1)
+                    {
+                        // C# compiler inverts this test, making the outer goto the conditional jmp.
+                        goto Equal;
+                    }
 
-                lengthToExamine -= sizeof(UIntPtr);
-                while ((byte*)lengthToExamine > (byte*)offset)
+                    // This becomes a conditional jmp foward to not favor it.
+                    goto NotEqual;
+                }
+                // Use Vector128.Size as Vector128<byte>.Count doesn't inline at R2R time
+                // https://github.com/dotnet/runtime/issues/32714
+                else if (length >= Vector128.Size)
                 {
-                    if (LoadUIntPtr(ref first, offset) != LoadUIntPtr(ref second, offset))
+                    Vector128<byte> vecResult;
+                    nuint offset = 0;
+                    nuint lengthToExamine = length - Vector128.Size;
+                    if (lengthToExamine != 0)
                     {
-                        goto NotEqual;
+                        do
+                        {
+                            // We use instrincs directly as .Equals calls .AsByte() which doesn't inline at R2R time
+                            // https://github.com/dotnet/runtime/issues/32714
+                            vecResult = Sse2.CompareEqual(LoadVector128(ref first, offset), LoadVector128(ref second, offset));
+                            if (Sse2.MoveMask(vecResult) != 0xFFFF)
+                            {
+                                goto NotEqual;
+                            }
+                            offset += Vector128.Size;
+                        } while (lengthToExamine > offset);
                     }
-                    offset += sizeof(UIntPtr);
-                }
-                return LoadUIntPtr(ref first, lengthToExamine) == LoadUIntPtr(ref second, lengthToExamine);
-            }
 
-            Debug.Assert((byte*)lengthToExamine < (byte*)sizeof(UIntPtr));
+                    // Do final compare as Vector128<byte>.Count from end rather than start
+                    Debug.Assert(lengthToExamine >= 0);
```

Comment:
Since _lengthToExamine_ is an unsigned data type this assertion will never trip so it can't detect errors in the logic.

Reply:
Perhaps `(lengthToExamine < length)`?

Reply:
Done